doc/whats_new/v0.24.rst

```diff
@@ -127,6 +127,9 @@ Changelog
   predictions for `est.transform(Y)` when the training data is single-target.
   :pr:`17095` by `Nicolas Hug`_.
 
+- |Fix| Increases the stability of :class:`cross_decomposition.CCA` :pr:`18746`
```

Comment:
specify that it's only for poorly conditioned matrices? So that not all users expect to have changes in the results

Reply:
The issue comes from:

https://github.com/scipy/scipy/blob/8e30f7797bd1ee442f4f1a25172e4402521c1e16/scipy/linalg/basic.py#L1385

when `pinv2` is called. On windows and a the pypi version of numpy the singlar values for `X` are :

```py
[0.7, 0.14, 9e-16]
```
which results in a rank of 3.

On other platforms, where the original test passes, the singular values are

```py
[0.7, 0.14, 1.4e-17]
```

which results in a rank of 2. This is by designed as stated [reference paper](https://stat.uw.edu/sites/default/files/files/reports/2000/tr371.pdf) on page 12.

This `Xk` is updated here:

https://github.com/scikit-learn/scikit-learn/blob/b5d63e34746ec273c1cbc5992a0477198a22f8be/sklearn/cross_decomposition/_pls.py#L263

Given, this I do not think its for poorly conditioned matrices. It is our usage of `pinv2` that was flaky.

Reply:
> when pinv2 is called. On windows and the pypi version of numpy the singular values for Xk are:

After which iteration? If this happens at the first iter, then this is indeed a problem of condition number I believe

Reply:
(by first iteration I mean the first component, or equivalently the first call to `_get_first_singular_vectors_power_method`)

Reply:
> After which iteration? If this happens at the first iter, then this is indeed a problem of condition number I believe

This happened on the second call to `_get_first_singular_vectors_power_method`.

Reply:
I think it would still be valuable to identify in which cases there's a stability improvement. Clearly, not all users will be impacted by this

Reply:
I think this happens randomly. I updated `test_scale_and_stability` to generate random `y`s and found that there were seeds that fail on my machine and on windows.

From a user point of view, for some datasets they may have gotten an incorrect model.

Reply:
Both random `y` *and* `X`? Because otherwise the problem might just be coming from `X`.

I still believe this is related to poorly conditioned matrices because what you pass to `_get_first_singular_vectors_power_method` at iteration i are the deflated matrices from iteration i - 1 (deflation = subtracting with a rank-1 matrix to obtain a (r - 1) rank matrix)

I'm quite certain that the name of the parameter to `pinv` (cond or rcond) is related to the *cond*ition number

Reply:
> Both random y and X? Because otherwise the problem might just be coming from X.

Updated with randomly generated X and y. The condition number will alway gets much bigger when the matrix becomes a r - 1 rank matrix.

When the rank becomes r - 1, `pinv2` by default is having trouble determining the rank of the matrix:

https://github.com/scipy/scipy/blob/8e30f7797bd1ee442f4f1a25172e4402521c1e16/scipy/linalg/basic.py#L1457-L1466

Setting `cond= 10 * eps` helps in this case.

Reply:
Specificly when `pinv2` thinks a `r-1` matrix is rank r, it divides by a really small singular value.